### DISCUSSION TITLE
Mesh with mesh collision sometimes does not generate contact manifold

### DISCUSSION THREAD

**Discussion by ghost** [bepu/bepuphysics2#206] [Q&A] [answered]
Mesh with mesh or mesh with box collision will not always generate contact manifolds. Currently I'm running V2.3, and cannot upgrade to 2.4 in the near future.  Is there another way to ensure there was a collision in the narrow phase?

**Comment by RossNordby** [comment, marked as answer]
Could you provide a repro? My guess is that this is related to one sided triangles. Triangle collisions and ray tests are one-sided; only tests which see the triangle as wound clockwise in right handed coordinates or counterclockwise in left handed coordinates will generate contacts.

To create double sided triangles, you can have two triangles with the same vertices but opposite winding. 2.4 does improve handling of this case a bit.

**Reply by ghost** [reply]
First of all thanks for your quick response. I tried double sided triangles but it would not work. So I extracted a scenario out of the program and built it in the Bepu v2.3 demo, So it is easier to find the problem, and have visualization of the shapes. The provided meshes are box shaped, but there are plenty scenario's that the program build a detailed concave shape, that's why i made the choice for going with a mesh.

[ReproMeshesWithNoContactManifold.zip](https://github.com/bepu/bepuphysics2/files/8349915/ReproMeshesWithNoContactManifold.zip)

**Reply by RossNordby** [reply]
Thanks for the repro. 

This appears to be related to a bug (or a family of bugs) that was fixed in 2.4. Edge-on collisions with triangles in 2.3 sometimes end up not generating contacts due to one sidedness and numerical issues that arose as collisions approached 90 degrees off face normal. This was more of an issue in triangle-triangle than some other pairs.

It is possible to backport the change with some work- the `TrianglePairTester` would be the start point. To retain compatibility with NS2.0, `Unsafe.SkipInit` would need to be removed, and some helpers in `TriangleWide`, `BundleIndexing`, and `ManifoldCandidateHelper` would need to be ported over.

**Reply by ghost** [reply]
Thanks for the quick response.

The proposed fix did the trick. It now creates manifold's when 2 meshes collides. Is it possible to do a pull request and implement this in 2.3? So other users that are still running on the .net framework can make use of this fix.

**Reply by RossNordby** [reply]
> Is it possible to do a pull request and implement this in 2.3? So other users that are still running on the .net framework can make use of this fix.

Sure- there's a 2.3.1 branch up now.
